src/p2p/net_node.inl

```diff
@@ -1704,4 +1749,30 @@ namespace nodetool
 
     return count > max_connections;
   }
+
+  template<class t_payload_net_handler>
+  bool node_server<t_payload_net_handler>::gray_peerlist_housekeeping()
+  {
+    peerlist_entry pe = AUTO_VAL_INIT(pe);
+
+    if (!m_peerlist.get_gray_peer_random(pe)) {
+        return false;
+    }
+
+    bool success = check_connection_and_handshake_with_peer(pe.adr, pe.last_seen);
+
+    if (!success) {
+      m_peerlist.remove_from_peer_gray(pe);
```

Comment:
The purpose is to connect to gray peers to evict non reachable peers or promote them if the connection is successful.

There could be non reachable peers because they are offline and can't foresee if they will come online again, or because a bad behaving peer sent non existent peers to fill the list.